lib/models/ActionQueueEntry.js

```diff
@@ -0,0 +1,393 @@
+     * it can be a simple attribute name, or a dot-separated path to a
+     * nested attribute. Missing parents are created automatically if
+     * needed.
+     * @param {object} value - attribute value, it can be any
+     * serializable JS object (string, number, array or object)
+     * @return {ActionQueueEntry} - this
+     */
+    setAttribute(attributePath, value) {
+        if (value === undefined) {
+            return this;
+        }
+        const attrPathElems = attributePath.split('.');
+        let parentAttr = this._actionAttributes;
+        for (let i = 0; i < attrPathElems.length - 1; ++i) {
+            const attrPathElem = attrPathElems[i];
+            if (parentAttr[attrPathElem] === undefined) {
+                parentAttr[attrPathElem] = {};
+            }
+            parentAttr = parentAttr[attrPathElem];
+        }
+        const lastPathElem = attrPathElems[attrPathElems.length - 1];
+        parentAttr[lastPathElem] = value;
+        return this;
+    }
+
+    /**
+     * Get the value of an action attribute
+     *
+     * @param {string} attributePath - path to the attribute to get -
+     * it can be a simple attribute name, or a dot-separated path to a
+     * nested attribute
+     * @return {object|undefined} - attribute value, or {undefined} if
+     * the attribute or any of its parents do not exist
+     */
+    getAttribute(attributePath) {
+        const attrPathElems = attributePath.split('.');
+        let attr = this._actionAttributes;
+        for (let i = 0; i < attrPathElems.length; ++i) {
+            const attrPathElem = attrPathElems[i];
+            attr = attr[attrPathElem];
+            if (attr === undefined) {
+                return undefined;
+            }
+        }
+        return attr;
+    }
+
+    /**
+     * Add one or more action attributes to be returned by
+     * getLogInfo(), in order to have them added to logs. Attributes
+     * to be logged but not present in the action object will not be
+     * logged.
+     *
+     * @param {object} attrMap - mapping of each attribute to be
+     * logged, where keys are log attribute names, and values are
+     * their matching action attribute path.
+     *   E.g. <code>{ actionFooBar: 'foo.bar' }</code>
+     * @return {ActionQueueEntry} - this
+     */
+    addLoggedAttributes(attrMap) {
+        Object.assign(this._loggedAttributes, attrMap);
+        return this;
+    }
+
+    getActionId() {
+        return this.getAttribute('actionId');
+    }
+
+    getActionType() {
+        return this.getAttribute('action');
+    }
+
+    /**
+     * Add one or more context attributes from a JS object, that will
+     * be stored in the "contextInfo" field of the action message and
+     * logged as part of getLogInfo().
+     *
+     * @param {object} contextObj - attributes to add to the action context
+     * @return {ActionQueueEntry} - this
+     */
+    addContext(contextObj) {
```

Comment:
just curious, is this being added for better logging? Will it serve purposes other than logging?

Reply:
Yes it will not only help for logging, but also to feed back some context to the action requester in the results topic messages. E.g., if transition policies request a data copy, in the response, the context can carry the necessary info about the original action, e.g. that it was a transition, or the original storage class for the purpose of conditional checks when updating the metadata. The idea is carrying state over to the next actions so that it's readily available when taking steps after the action is done.